File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

bool solution(vector<string> phone_book) {
    unordered_map<string, bool> prefixMap;

    for (string number : phone_book) {
        for (int i = 1; i < number.length(); i++) {
            prefixMap[number.substr(0, i)] = true;
        }
    }

    for (string number : phone_book) {
        if (prefixMap.find(number) != prefixMap.end()) {
            return false;
        }
    }
    return true;
}

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.HashMap;

class Solution {

	public boolean solution(String[] phone_book) {
		HashMap<String, Boolean> map = new HashMap<>();
		for (String number : phone_book) {
			map.put(number, true);
		}

		for (String number : phone_book) {
			for (int i = 1; i < number.length(); i++) {
				if (map.containsKey(number.substring(0, i))) {
					return false;
				}
			}
		}
		return true;
	}
}
